skimage/filters/_sparse.py

```diff
@@ -23,13 +22,59 @@ def _validate_window_size(axis_sizes):
             raise ValueError(msg)
 
 
-def _to_np_mode(mode):
-    """Convert padding modes from `ndi.correlate` to `np.pad`."""
-    mode_translation_dict = dict(nearest='edge', reflect='symmetric',
-                                 mirror='reflect')
-    if mode in mode_translation_dict:
-        mode = mode_translation_dict[mode]
-    return mode
+def _get_view(padded, kernel_shape, idx, val):
+    """Get a view into `padded` that is offset by `idx` and scaled by `val`.
+
+    If `padded` was created by padding the original image by `kernel_shape` as
+    in _mean_std, then the view created here will match the size of the
+    original image.
+    """
+    sl_shift = tuple([slice(c, s - (w_ - 1 - c))
+                      for c, w_, s in zip(idx, kernel_shape, padded.shape)])
+    v = padded[sl_shift]
+    if val == 1:
+        return v
+    elif val == -1:
+        return -v
+    return val * v
+
+
+def _correlate_sparse(image, kernel_shape, kernel_indices_and_values):
+    """Perform correlation with a sparse kernel.
+
+    Parameters
+    ----------
+    image : ndarray
+        The (prepadded) image to be correlated.
+    kernel_shape : tuple of int
+        The shape of the sparse filter kernel.
+    kernel_indices_and_values : list of 2-tuples
+        This is a list of 2-tuples with length equal to the number of nonzero
+        kernel entries. The first element of the tuple is the coordinate within
+        `kernel_shape` and the second element is the kernel value at that
+        coordinate.
+
+    Returns
+    -------
+    out : ndarray
+        The filtered image.
+
+    Notes
+    -----
+    This function only returns results for the 'valid' region of the
+    convolution, and thus `out` will be smaller than `image` by an amount
+    equal to the kernel size along each axis.
+    """
+    idx, val = kernel_indices_and_values[0]
+    # implementation assumes this corner is first in kernel_indices_in_values
+    assert tuple(idx) == (0, ) * image.ndim
+    out = _get_view(image, kernel_shape, idx, val)
+    if not out.flags.owndata:
```

Comment:
I am fine with that and don't think it will substantially reduce performance, but given the reliance on `scipy.ndimage` elsewhere in the library for many things related to filtering/morphology/labeling (as well as usage of FFTs or sparse matrices in various places), we will already not be able to realistically do many things purely with the Array API. I think NumPy+SciPy, Dask+dask-image and CuPy are the only implementations that provide most of the needed features currently.

Reply:
Well, dask arrays don't have .flags either...

```python
In [1]: arr = da.random.random((5,))
In [2]: arr.flags
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
<ipython-input-2-c469d1119f64> in <module>
----> 1 arr.flags

AttributeError: 'Array' object has no attribute 'flags'
```